File: OmniScript/py/OmniScript/omniscript/countrystatistic.py
"""CountryStatistic class.

The :class:`Capture <omniscript.capture.Capture>` function
:func:`get_call_stats() <omniscript.capture.Capture.get_call_stats>`
returns a CountryStatistic objects.
"""
#Copyright (c) WildPackets, Inc. 2015. All rights reserved.

import os
import sys
import logging
import time

import omniscript

from peektime import PeekTime


class CountryStatistic(object):
    """The CountryStatistic class has the statistic of the country the
    packet originated or was received in.
    """

    bytes_from = 0
    """The number of bytes received from the Country."""

    bytes_to = 0
    """The number of bytes sent to the Country."""

    country_code = ""
    """The 2 letter ISO Country Code string, Unknown, Private or Multicast."""

    country_name = ""
    """The name of the country, Unknown, Private or Multicast."""

    first_time_from = None
    """The first time, as
    :class:`PeekTime <omniscript.peektime.PeekTime>`,
    a packet was received from the Country."""

    first_time_to = None
    """The first time, as
    :class:`PeekTime <omniscript.peektime.PeekTime>`,
    a packet was sent to the Country.
    """

    last_time_from = None
    """The last time, as
    :class:`PeekTime <omniscript.peektime.PeekTime>`,
    a packet was received from the Country."""

    last_time_to = None
    """The last time, as
    :class:`PeekTime <omniscript.peektime.PeekTime>`,
    a packet was sent to the Country."""

    packets_from = 0
    """The number of packets received from the Country."""

    packets_to = 0
    """The number of packets sent to the Country."""

    def __init__(self, stream):
        self.bytes_from = CountryStatistic.bytes_from
        self.bytes_to = CountryStatistic.bytes_to
        self.country_code = CountryStatistic.country_code
        self.country_name = CountryStatistic.country_name
        self.first_time_from = CountryStatistic.first_time_from
        self.first_time_to = CountryStatistic.first_time_to
        self.last_time_from = CountryStatistic.last_time_from
        self.last_time_to = CountryStatistic.last_time_to
        self.packets_from = CountryStatistic.packets_from
        self.packets_to = CountryStatistic.packets_to
        if stream is not None:
            self._load(stream)

    def _load(self, stream):
        code = stream.read_string(2)
        if ord(code[0]) == 0:
            if ord(code[1]) == 0:
                code = 'Unknown'
            elif ord(code[1]) == 1:
                code = 'Private'
            elif ord(code[1]) == 2:
                code = 'Multicast'
        self.country_code = code
        stream.seek(6, os.SEEK_CUR) # padding
        self.first_time_from = PeekTime(stream.read_ulong())
        self.last_time_from = PeekTime(stream.read_ulong())
        self.packets_from = stream.read_ulong()
        self.bytes_from = stream.read_ulong()
        self.first_time_to = PeekTime(stream.read_ulong())
        self.last_time_to = PeekTime(stream.read_ulong())
        self.packets_to = stream.read_ulong()
        self.bytes_to = stream.read_ulong()
        co_names = omniscript.get_country_code_names()
        if code in co_names:
            self.country_name = co_names[code]
        
    def get(self):
        return self

File: OmniScript/cs/OmniScript/CountryStatistic.cs
// =============================================================================
// <copyright file="Adapter.cs" company="LiveAction, Inc.">
//  Copyright (c) 2018-2021 Savvius, Inc. All rights reserved.
//  Copyright (c) 2015-2018 Savvius, Inc. All rights reserved.
// </copyright>
// =============================================================================

namespace Savvius.Omni.OmniScript
{
    using System;
    using System.Collections.Generic;

    public class CountryStatistic
    {
        public String CountryCode;
        public String CountryName;
        public PeekTime FirstTimeFrom;
        public PeekTime LastTimeFrom;
        public ulong PacketsFrom;
        public ulong BytesFrom;
        public PeekTime FirstTimeTo;
        public PeekTime LastTimeTo;
        public ulong PacketsTo;
        public ulong BytesTo;

        public CountryStatistic(ReadStream stream, MediaSpec spec)
        {
            this.CountryCode = "";
            this.CountryName = "";
            this.FirstTimeFrom = null;
            this.LastTimeFrom = null;
            this.PacketsFrom = 0;
            this.BytesFrom = 0;
            this.FirstTimeTo = null;
            this.LastTimeTo = null;
            this.PacketsTo = 0;
            this.BytesTo = 0;

            this.Load(stream);
        }

        private void Load(ReadStream stream)
        {
            if (stream == null) return;

            this.CountryCode = stream.ReadString(2);
            stream.Read(6); // Padding.
            this.FirstTimeFrom = new PeekTime(stream.ReadULong());
            this.LastTimeFrom = new PeekTime(stream.ReadULong());
            this.PacketsFrom = stream.ReadULong();
            this.BytesFrom = stream.ReadULong();
            this.FirstTimeTo = new PeekTime(stream.ReadULong());
            this.LastTimeTo = new PeekTime(stream.ReadULong());
            this.PacketsTo = stream.ReadULong();
            this.BytesTo = stream.ReadULong();

            this.CountryName = OmniScript.CoCodeNameMap[this.CountryCode];
        }

        public String Report(char seperator, bool newline)
        {
            uint index = 1;
            String line = String.Format("{1}{0}{2}{0}{3}{0}{4}",
                seperator,
                index,
                this.CountryCode,
                this.BytesFrom,
                this.PacketsFrom);
            if (newline)
            {
                line += Environment.NewLine;
            }
            return line;
        }
    }

    public class CountryStatistics : List<ProtocolStatistic>
    {
    }
}
